From cce0a7663373b3b26837d2733fef170174cd61be Mon Sep 17 00:00:00 2001
From: ywan171 <yi.a.wang@intel.com>
Date: Tue, 30 Jul 2013 22:38:43 +0800
Subject: ATSParser: fix if continuity_counter is not consecutive, corruption issue

BZ: 110261

fix if continuity_counter is not consecutive,

ATSParser will drop inappropriate packet and cause corruption issue

Category: aosp improvement
Domain: Video.Media-streaming
Origin: internal
Upstream-Candidate: yes

Change-Id: Ibf779e9a51f3807d1f5cf6a821010796793a5e1f
Orig-Change-Id: I9830e4aa8fa2c6ffaa96f0bfedcecd4bf2d73b84
Signed-off-by: ywan171 <yi.a.wang@intel.com>
Signed-off-by: chuzhoux <chux.zhou@intel.com>
---
 media/libstagefright/mpeg2ts/ATSParser.cpp |    9 ++++-----
 1 files changed, 4 insertions(+), 5 deletions(-)

diff --git a/media/libstagefright/mpeg2ts/ATSParser.cpp b/media/libstagefright/mpeg2ts/ATSParser.cpp
index 3fb39e1..25bc8bf 100644
--- a/media/libstagefright/mpeg2ts/ATSParser.cpp
+++ b/media/libstagefright/mpeg2ts/ATSParser.cpp
@@ -532,10 +532,11 @@ status_t ATSParser::Stream::parse(
 
     if (mExpectedContinuityCounter >= 0
             && (unsigned)mExpectedContinuityCounter != continuity_counter) {
-        ALOGI("discontinuity on stream pid 0x%04x", mElementaryPID);
-
+        ALOGI("discontinuity on stream pid 0x%04x,expected = %d, got = %d",
+            mElementaryPID, mExpectedContinuityCounter, continuity_counter);
         mPayloadStarted = false;
-        mBuffer->setRange(0, 0);
+        // check if there's a complete PES in current payload buffer
+        flush();
         mExpectedContinuityCounter = -1;
 
 #if 0
@@ -547,8 +548,6 @@ status_t ATSParser::Stream::parse(
             mQueue->clear(true /* clearFormat */);
         }
 #endif
-
-        return OK;
     }
 
     mExpectedContinuityCounter = (continuity_counter + 1) & 0x0f;
-- 
1.7.4.1

